File: runtime/worlds/rpg.py
"""Grid RPG world: a turn-based dungeon, as a held-out decision probe.

Plan: `.claude/plans/rpg-demo-app.md`. Unlike every other world here, the
tools are not CRUD — the rules live in `runtime/engines/rpg.js` and run
inside the sandbox through the `engine` impl op, so the browser demo and the
headless suite execute identical dynamics. This module owns the schema, the
scenarios, and *perception*: turning a state into the request text and
constants the model sees. It never mutates the game.

Reserved in `data/holdout/reserved.json` — never in training data.

State shape, beyond the standard `entities`/`outbox`/`payments`:
  map              {width, height, rows[]}   '#' wall, '.' floor, 'E' stairs,
                                             'D' the door's tile
  turn             int, incremented by end_turn
  status           playing | won | dead
  log              events from the last turn (rendered as "Last turn:")
  memory           entity ids seen at least once (written by the caller from
                   Observation.memory)
  turn_budget      actions allowed per turn (default 3)
  actions_this_turn spent so far this turn; end_turn resets it
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

NOW = 1_760_000_000

WALL = "#"
EXIT = "E"
DIRECTIONS = ["north", "south", "east", "west"]
DIR_DESC = {
    "north": "direction: north (up on the map, y-1)",
    "south": "direction: south (down on the map, y+1)",
    "east": "direction: east (right on the map, x+1)",
    "west": "direction: west (left on the map, x-1)",
}

WORLD = {
    "name": "rpg",
    "now": NOW,
    "entities": {
        "player": {
            "id": "ID:player",
            "x": "INT",
            "y": "INT",
            "hp": "INT",
            "max_hp": "INT",
            "attack": "INT",
        },
        "enemy": {
            "id": "ID:enemy",
            "kind": "STR",
            "x": "INT",
            "y": "INT",
            "hp": "INT",
            "attack": "INT",
        },
        "item": {
            "id": "ID:item",
            "kind": "STR",     # key | potion
            "x": "INT",
            "y": "INT",
            "held": "BOOL",
        },
        "door": {
            "id": "ID:door",
            "x": "INT",
            "y": "INT",
            "locked": "BOOL",
            "open": "BOOL",
        },
    },
    "enums": {("item", "kind"): ["key", "potion"]},
    "tools": [
        {
            "name": "move",
            "desc": "Walk one tile in a direction. Fails if a wall, a closed "
                    "door, or an enemy is in the way.",
            "params": [{"name": "direction", "type": "STR",
                        "desc": "which way to walk: north, south, east or west"}],
            "returns": "OBJ:player",
            "effects": ["WRITE"],
            "impl": {"op": "engine", "module": "rpg", "fn": "move"},
        },
        {
            "name": "attack",
            "desc": "Strike an enemy standing on a tile next to you. Fails if "
                    "it is further away than one tile.",
            "params": [{"name": "target", "type": "ID:enemy",
                        "desc": "the enemy to hit", "field": ["enemy", "id"]}],
            "returns": "OBJ:enemy",
            "effects": ["WRITE"],
            "impl": {"op": "engine", "module": "rpg", "fn": "attack"},
        },
        {
            "name": "pick_up",
            "desc": "Take an item lying on the tile you are standing on. Fails "
                    "if the item is anywhere else.",
            "params": [{"name": "item", "type": "ID:item",
                        "desc": "the item on the floor here",
                        "field": ["item", "id"]}],
            "returns": "OBJ:item",
            "effects": ["WRITE"],
            "impl": {"op": "engine", "module": "rpg", "fn": "pick_up"},
        },
        {
            "name": "use_item",
            "desc": "Use an item you are carrying. Drinking a potion restores "
                    "health. A key is not used this way — open its door with "
                    "interact instead.",
            "params": [
                {"name": "item", "type": "ID:item",
                 "desc": "the carried item to use", "field": ["item", "id"]},
                {"name": "target", "type": "ID:enemy", "required": False,
                 "desc": "enemy to use it on, when the item needs one",
                 "field": ["enemy", "id"]},
            ],
            "returns": "OBJ:player",
            "effects": ["WRITE"],
            "impl": {"op": "engine", "module": "rpg", "fn": "use_item"},
        },
        {
            "name": "interact",
            "desc": "Open a door on a tile next to you. A locked door opens "
                    "only if you are carrying its key, which is used up.",
            "params": [{"name": "target", "type": "ID:door",
                        "desc": "the door to open", "field": ["door", "id"]}],
            "returns": "OBJ:door",
            "effects": ["WRITE"],
            "impl": {"op": "engine", "module": "rpg", "fn": "interact"},
        },
    ],
    # runtime/sandbox.js runs this after the program ends: the enemy phase.
    "post_hook": {"module": "rpg", "fn": "end_turn"},
}


# --------------------------------------------------------------- scenarios

KEEP_ROWS = [
    "############",
    "#......#...#",
    "#..g...#.E.#",
    "#......D...#",
    "#......#####",
    "#.p....#..k#",
    "#..........#",
    "#@.....#.g.#",
    "############",
]

SCENARIOS = {
    "keep": {
        "quest": "Find the stairs down. They are in the north-east, behind a "
                 "locked door; the key is somewhere off to the east.",
        "vision": 2,
        "max_turns": 40,
        "rows": KEEP_ROWS,
        "player": {"x": 1, "y": 7, "hp": 10, "max_hp": 10, "attack": 3},
        "enemies": [
            {"kind": "goblin", "x": 3, "y": 2, "hp": 3, "attack": 2},
            {"kind": "goblin", "x": 9, "y": 7, "hp": 3, "attack": 2},
        ],
        "items": [
            {"kind": "potion", "x": 2, "y": 5},
            {"kind": "key", "x": 10, "y": 5},
        ],
        "doors": [{"x": 7, "y": 3, "locked": True, "open": False}],
    },
}


def new_state(scenario: str = "keep") -> dict:
    """A fresh game. The map glyphs are display only — walls are the one
    thing they decide; doors and the exit are entities/tiles the engine
    reads."""
    sc = SCENARIOS[scenario]
    rows = list(sc["rows"])
    return {
        "entities": {
            "player": [dict(id="player_1", **sc["player"])],
            "enemy": [dict(id=f"enemy_{i + 1}", **e)
                      for i, e in enumerate(sc["enemies"])],
            "item": [dict(id=f"item_{i + 1}", held=False, **it)
                     for i, it in enumerate(sc["items"])],
            "door": [dict(id=f"door_{i + 1}", **d)
                     for i, d in enumerate(sc["doors"])],
        },
        "outbox": [],
        "payments": [],
        "map": {"width": len(rows[0]), "height": len(rows), "rows": rows},
        "scenario": scenario,
        "quest": sc["quest"],
        "vision": sc["vision"],
        "max_turns": sc["max_turns"],
        "turn": 0,
        "status": "playing",
        "log": [],
        "memory": [],
        "turn_budget": 3,
        "actions_this_turn": 0,
    }


WORLD["default_state"] = new_state()


# -------------------------------------------------------------- perception

@dataclass
class Observation:
    """What one turn looks like to the model, plus the structured form the UI
    draws its fog from. `memory` is the updated seen-set: the caller writes it
    back to state["memory"] (perception is otherwise pure)."""
    request: str
    constants: List[dict]
    window: List[str]
    nearby: List[dict]
    memory: List[str]


def _player(state: dict) -> dict:
    return state["entities"]["player"][0]


def _tile(state: dict, x: int, y: int) -> str:
    m = state["map"]
    if x < 0 or y < 0 or x >= m["width"] or y >= m["height"]:
        return WALL
    row = m["rows"][y]
    return row[x] if x < len(row) else WALL


def _offset_words(dx: int, dy: int) -> str:
    """(dx, dy) in tiles -> "2 east 1 south"; (0,0) -> "on your tile"."""
    parts = []
    if dx:
        parts.append(f"{abs(dx)} {'east' if dx > 0 else 'west'}")
    if dy:
        parts.append(f"{abs(dy)} {'south' if dy > 0 else 'north'}")
    return " ".join(parts) if parts else "on your tile"


def _visible(state: dict, vision: int) -> dict:
    """Everything inside the square window, by kind. Vision is a plain radius:
    no line-of-sight, so a wall does not hide what is behind it (deliberately
    simple — the probe is decision-making, not raycasting)."""
    p = _player(state)
    ents = state["entities"]

    def near(e):
        return abs(e["x"] - p["x"]) <= vision and abs(e["y"] - p["y"]) <= vision

    return {
        "enemy": [e for e in ents.get("enemy", []) if e["hp"] > 0 and near(e)],
        "item": [i for i in ents.get("item", [])
                 if not i.get("held") and near(i)],
        "door": [d for d in ents.get("door", []) if near(d)],
        "held": [i for i in ents.get("item", []) if i.get("held")],
    }


def _glyph(state: dict, x: int, y: int, vis: dict) -> str:
    p = _player(state)
    if x == p["x"] and y == p["y"]:
        return "@"
    for e in vis["enemy"]:
        if e["x"] == x and e["y"] == y:
            return e["kind"][0].lower()
    for i in vis["item"]:
        if i["x"] == x and i["y"] == y:
            return "k" if i["kind"] == "key" else "p"
    for d in vis["door"]:
        if d["x"] == x and d["y"] == y:
            return "/" if d["open"] else "D"
    t = _tile(state, x, y)
    return t if t in (WALL, EXIT) else "."


def observe(state: dict, vision: Optional[int] = None) -> Observation:
    """Render one turn: the request text the model reads and the constants it
    may name. Only things it can see (or carries) become constants — the
    kanban lesson (`relevant_cards`, server/dev_server.py): dumping every
    entity made the tuned model pick the wrong one. Directions come first so
    the oracle's $0..$3 are stable."""
    p = _player(state)
    vision = state.get("vision", 2) if vision is None else vision
    vis = _visible(state, vision)

    window = []
    for y in range(p["y"] - vision, p["y"] + vision + 1):
        window.append("".join(_glyph(state, x, y, vis)
                              for x in range(p["x"] - vision, p["x"] + vision + 1)))

    constants: List[dict] = [
        {"type": "STR", "value": d, "desc": DIR_DESC[d]} for d in DIRECTIONS
    ]
    nearby: List[dict] = []

    for e in sorted(vis["enemy"], key=lambda e: e["id"]):
        dx, dy = e["x"] - p["x"], e["y"] - p["y"]
        adj = abs(dx) + abs(dy) == 1
        where = _offset_words(dx, dy)
        desc = (f"{e['kind']} {where}, {'adjacent' if adj else 'not adjacent'}"
                f", {e['hp']} HP")
        constants.append({"type": "ID:enemy", "value": e["id"], "desc": desc})
        nearby.append({"id": e["id"], "kind": e["kind"], "type": "enemy",
                       "dx": dx, "dy": dy, "adjacent": adj, "hp": e["hp"]})

    for i in sorted(vis["item"], key=lambda i: i["id"]):
        dx, dy = i["x"] - p["x"], i["y"] - p["y"]
        here = dx == 0 and dy == 0
        desc = (f"{i['kind']} on the floor "
                f"{'here, on your tile' if here else _offset_words(dx, dy)}")
        constants.append({"type": "ID:item", "value": i["id"], "desc": desc})
        nearby.append({"id": i["id"], "kind": i["kind"], "type": "item",
                       "dx": dx, "dy": dy, "here": here})

    for d in sorted(vis["door"], key=lambda d: d["id"]):
        dx, dy = d["x"] - p["x"], d["y"] - p["y"]
        adj = abs(dx) + abs(dy) == 1
        state_word = ("open" if d["open"]
                      else "locked" if d["locked"] else "shut")
        desc = (f"{state_word} door {_offset_words(dx, dy)}, "
                f"{'adjacent' if adj else 'not adjacent'}")
        constants.append({"type": "ID:door", "value": d["id"], "desc": desc})
        nearby.append({"id": d["id"], "type": "door", "dx": dx, "dy": dy,
                       "adjacent": adj, "locked": d["locked"], "open": d["open"]})

    for i in sorted(vis["held"], key=lambda i: i["id"]):
        constants.append({"type": "ID:item", "value": i["id"],
                          "desc": f"{i['kind']} you are carrying"})

    memory = list(state.get("memory", []))
    for group in ("enemy", "item", "door"):
        for e in vis[group]:
            if e["id"] not in memory:
                memory.append(e["id"])

    visible_ids = {e["id"] for g in ("enemy", "item", "door") for e in vis[g]}
    remembered = _remembered_line(state, memory, visible_ids)
    carrying = ", ".join(sorted(i["kind"] for i in vis["held"])) or "nothing"
    nearby_line = "; ".join(
        f"{n.get('kind', 'door')} {n['id']} "
        f"{'on your tile' if n.get('here') else _offset_words(n['dx'], n['dy'])}"
        for n in nearby) or "nothing"
    last = "; ".join(state.get("log") or []) or "nothing yet"

    request = (
        f"Dungeon crawl, turn {state.get('turn', 0)}. {state.get('quest', '')} "
        f"You are at ({p['x']},{p['y']}) with {p['hp']}/{p['max_hp']} HP, "
        f"carrying {carrying}.\n"
        f"What you can see ({2 * vision + 1}x{2 * vision + 1} tiles around you, "
        f"north is up, you are @ in the middle): "
        f"# wall, . floor, E stairs down, D locked door, / open door, "
        f"g goblin, k key, p potion.\n"
        + "\n".join(window) + "\n"
        f"Nearby: {nearby_line}.\n"
        f"Seen earlier, not visible now: {remembered}.\n"
        f"Last turn: {last}.\n"
        f"Choose up to {state.get('turn_budget', 3)} actions for this turn."
    )
    return Observation(request=request, constants=constants, window=window,
                       nearby=nearby, memory=memory)


def _remembered_line(state: dict, memory: List[str],
                     visible_ids: set) -> str:
    ents = state["entities"]
    by_id = {}
    for group in ("enemy", "item", "door"):
        for e in ents.get(group, []):
            by_id[e["id"]] = (group, e)
    out = []
    p = _player(state)
    for eid in memory:
        if eid in visible_ids or eid not in by_id:
            continue
        group, e = by_id[eid]
        if group == "item" and e.get("held"):
            continue
        if group == "enemy" and e["hp"] <= 0:
            continue
        name = e.get("kind", "door")
        out.append(f"{name} {_offset_words(e['x'] - p['x'], e['y'] - p['y'])}")
    return "; ".join(out) if out else "nothing"


def outcome(state: dict) -> dict:
    """Terminal predicate + the progress funnel the suite scores episodes on
    (goal_success is state equality against one reference — wrong shape for a
    game with many winning play-throughs)."""
    p = _player(state)
    items = state["entities"].get("item", [])
    doors = state["entities"].get("door", [])
    return {
        "status": state.get("status", "playing"),
        "won": state.get("status") == "won",
        "dead": state.get("status") == "dead",
        "turn": state.get("turn", 0),
        "hp": p["hp"],
        # the key is consumed by unlocking, so "taken" means picked up or used
        "key_taken": not any(i["kind"] == "key" and not i.get("held")
                             for i in items),
        "door_opened": all(d["open"] for d in doors) if doors else False,
        "enemies_left": len([e for e in state["entities"].get("enemy", [])
                             if e["hp"] > 0]),
    }

File: runtime/engines/rpg.js
"use strict";
/*
 * RPG world rules (plan .claude/plans/rpg-demo-app.md).
 *
 * Loaded by runtime/sandbox.js's `engine` impl op. Every tool the model can
 * call lands in one of the exported functions; `end_turn` is the enemy phase,
 * run once per turn by the sandbox's post_hook after the program finishes.
 *
 * Contract with the sandbox:
 *   fn(state, params, rt) -> result
 *   rt = { now, ToolError, clone }
 * Functions mutate `state` in place and push a line onto `state.log`.
 *
 * VALIDATE THEN MUTATE. The sandbox returns the (mutated) state even when a
 * tool throws (runtime/sandbox.js catch), so a function that mutates before
 * discovering the action is illegal would leave a half-applied action in a
 * state the harness keeps. Every check happens before the first write.
 *
 * Deterministic: no RNG anywhere. Same program + same state => same result,
 * which is what makes paired cross-model episodes comparable.
 */

const DIRS = {
  north: [0, -1],
  south: [0, 1],
  east: [1, 0],
  west: [-1, 0],
};

const WALL = "#";
const EXIT = "E";
const AGGRO = 3;          // Manhattan distance at which an enemy engages
const POTION_HEAL = 5;
const TURN_BUDGET_DEFAULT = 3;

// ---------------------------------------------------------------- helpers

function player(state) {
  const list = (state.entities && state.entities.player) || [];
  return list[0];
}

function enemies(state) {
  return (state.entities && state.entities.enemy) || [];
}

function items(state) {
  return (state.entities && state.entities.item) || [];
}

function doors(state) {
  return (state.entities && state.entities.door) || [];
}

function tileAt(state, x, y) {
  const map = state.map || { width: 0, height: 0, rows: [] };
  if (x < 0 || y < 0 || x >= map.width || y >= map.height) return WALL;
  const row = map.rows[y] || "";
  return row[x] || WALL;
}

function enemyAt(state, x, y) {
  return enemies(state).find((e) => e.x === x && e.y === y && e.hp > 0);
}

function doorAt(state, x, y) {
  return doors(state).find((d) => d.x === x && d.y === y);
}

function adjacent(a, b) {
  return Math.abs(a.x - b.x) + Math.abs(a.y - b.y) === 1;
}

function log(state, text) {
  if (!Array.isArray(state.log)) state.log = [];
  state.log.push(text);
}

/** Every action goes through here first: the game must be running, the
 *  player alive, and the turn's action budget not yet spent.
 *
 *  It also clears last turn's events. `log` is what the next observation
 *  renders as "Last turn:", so it must hold exactly one turn: the player's
 *  actions plus the enemy phase that answered them. Clearing here rather
 *  than in end_turn is what keeps the enemy's moves attached to the turn
 *  they belong to — clearing at the end would drop them before they were
 *  ever shown. (Without this the log grew for the whole episode and the
 *  prompt filled up with history: caught 2026-09-04 by reading turn 7 of an
 *  oracle run.) */
function beginAction(state, rt) {
  const p = player(state);
  if (!p) throw new rt.ToolError("NOT_FOUND", "no player in this world");
  if (!(state.actions_this_turn > 0)) state.log = [];
  if (state.status === "won") {
    throw new rt.ToolError("INVALID_ARGUMENT", "the dungeon is already cleared");
  }
  if (state.status === "dead" || p.hp <= 0) {
    throw new rt.ToolError("INVALID_ARGUMENT", "you are dead");
  }
  const budget = state.turn_budget === undefined
    ? TURN_BUDGET_DEFAULT : state.turn_budget;
  const spent = state.actions_this_turn || 0;
  if (spent >= budget) {
    throw new rt.ToolError("RATE_LIMITED",
      `turn budget spent (${budget} actions per turn)`);
  }
  return p;
}

/** Charged for every action that reaches the mutation stage AND for every
 *  action rejected as illegal — otherwise a TRY-wrapped program could probe
 *  the map for free. Callers invoke it after validation, before mutation;
 *  the throw paths call it explicitly. */
function spend(state) {
  state.actions_this_turn = (state.actions_this_turn || 0) + 1;
}

function fail(state, rt, code, message) {
  spend(state);
  log(state, `failed: ${message}`);
  return new rt.ToolError(code, message);
}

/** Walkable for the player: floor/exit tile, no living enemy, no closed door. */
function blockedFor(state, x, y) {
  const t = tileAt(state, x, y);
  if (t === WALL) return "a wall";
  const door = doorAt(state, x, y);
  if (door && !door.open) return door.locked ? "a locked door" : "a closed door";
  if (enemyAt(state, x, y)) return "an enemy";
  return null;
}

// ------------------------------------------------------------------ tools

function move(state, params, rt) {
  const p = beginAction(state, rt);
  const dir = String(params[0] || "").toLowerCase();
  if (!(dir in DIRS)) {
    throw fail(state, rt, "INVALID_ARGUMENT",
      `${params[0]} is not a direction (north, south, east, west)`);
  }
  const [dx, dy] = DIRS[dir];
  const nx = p.x + dx;
  const ny = p.y + dy;
  const blocker = blockedFor(state, nx, ny);
  if (blocker) {
    throw fail(state, rt, "INVALID_ARGUMENT", `cannot move ${dir}: ${blocker}`);
  }
  spend(state);
  p.x = nx;
  p.y = ny;
  log(state, `moved ${dir}`);
  if (tileAt(state, nx, ny) === EXIT) {
    state.status = "won";
    log(state, "reached the stairs down — the dungeon is cleared");
  }
  return rt.clone(p);
}

function attack(state, params, rt) {
  const p = beginAction(state, rt);
  const target = enemies(state).find((e) => e.id === params[0]);
  if (!target || target.hp <= 0) {
    throw fail(state, rt, "NOT_FOUND", `no enemy ${params[0]} here`);
  }
  if (!adjacent(p, target)) {
    throw fail(state, rt, "INVALID_ARGUMENT",
      `${target.kind} is not adjacent — step next to it first`);
  }
  spend(state);
  target.hp -= p.attack;
  if (target.hp <= 0) {
    target.hp = 0;
    state.entities.enemy = enemies(state).filter((e) => e.id !== target.id);
    log(state, `killed the ${target.kind}`);
  } else {
    log(state, `hit the ${target.kind} for ${p.attack} (${target.hp} HP left)`);
  }
  return rt.clone(target);
}

function pick_up(state, params, rt) {
  const p = beginAction(state, rt);
  const item = items(state).find((i) => i.id === params[0]);
  if (!item || item.held) {
    throw fail(state, rt, "NOT_FOUND", `no item ${params[0]} on the floor`);
  }
  if (item.x !== p.x || item.y !== p.y) {
    throw fail(state, rt, "INVALID_ARGUMENT",
      `the ${item.kind} is not on your tile — walk onto it first`);
  }
  spend(state);
  item.held = true;
  item.x = -1;
  item.y = -1;
  log(state, `picked up the ${item.kind}`);
  return rt.clone(item);
}

function use_item(state, params, rt) {
  const p = beginAction(state, rt);
  const item = items(state).find((i) => i.id === params[0]);
  if (!item) throw fail(state, rt, "NOT_FOUND", `no item ${params[0]}`);
  if (!item.held) {
    throw fail(state, rt, "INVALID_ARGUMENT",
      `you are not carrying the ${item.kind} — pick it up first`);
  }
  if (item.kind === "key") {
    throw fail(state, rt, "INVALID_ARGUMENT",
      "a key is not used on its own — interact with the locked door");
  }
  if (item.kind !== "potion") {
    throw fail(state, rt, "INVALID_ARGUMENT", `cannot use a ${item.kind}`);
  }
  spend(state);
  const healed = Math.min(POTION_HEAL, p.max_hp - p.hp);
  p.hp += healed;
  state.entities.item = items(state).filter((i) => i.id !== item.id);
  log(state, `drank the potion, healed ${healed} (${p.hp}/${p.max_hp} HP)`);
  return rt.clone(p);
}

function interact(state, params, rt) {
  const p = beginAction(state, rt);
  const door = doors(state).find((d) => d.id === params[0]);
  if (!door) throw fail(state, rt, "NOT_FOUND", `no door ${params[0]}`);
  if (!adjacent(p, door)) {
    throw fail(state, rt, "INVALID_ARGUMENT",
      "the door is not adjacent — step next to it first");
  }
  if (door.open) {
    throw fail(state, rt, "INVALID_ARGUMENT", "that door is already open");
  }
  if (door.locked) {
    const key = items(state).find((i) => i.kind === "key" && i.held);
    if (!key) {
      throw fail(state, rt, "INVALID_ARGUMENT",
        "the door is locked and you have no key");
    }
    spend(state);
    door.locked = false;
    door.open = true;
    state.entities.item = items(state).filter((i) => i.id !== key.id);
    log(state, "unlocked the door with the key and opened it");
    return rt.clone(door);
  }
  spend(state);
  door.open = true;
  log(state, "opened the door");
  return rt.clone(door);
}

// -------------------------------------------------------------- turn end

/** The enemy phase, run once per turn by the sandbox post_hook after the
 *  program ends (see the plan: acting after every call would invalidate the
 *  snapshot the model planned against). Also rolls the turn counter and
 *  clears the per-turn action budget. */
function end_turn(state) {
  const p = player(state);
  if (!p) return state;
  if (state.status === "playing" && p.hp > 0) {
    for (const e of enemies(state)) {
      if (e.hp <= 0) continue;
      const dist = Math.abs(e.x - p.x) + Math.abs(e.y - p.y);
      if (dist > AGGRO) continue;
      if (dist === 1) {
        p.hp -= e.attack;
        log(state, `the ${e.kind} hits you for ${e.attack}`);
        if (p.hp <= 0) {
          p.hp = 0;
          state.status = "dead";
          log(state, "you died");
          break;
        }
        continue;
      }
      // step one tile toward the player: larger axis first, then the other
      const dx = p.x - e.x;
      const dy = p.y - e.y;
      const tries = Math.abs(dx) >= Math.abs(dy)
        ? [[Math.sign(dx), 0], [0, Math.sign(dy)]]
        : [[0, Math.sign(dy)], [Math.sign(dx), 0]];
      for (const [sx, sy] of tries) {
        if (sx === 0 && sy === 0) continue;
        const nx = e.x + sx;
        const ny = e.y + sy;
        if (tileAt(state, nx, ny) === WALL) continue;
        const door = doorAt(state, nx, ny);
        if (door && !door.open) continue;
        if (enemyAt(state, nx, ny)) continue;
        if (nx === p.x && ny === p.y) continue;
        e.x = nx;
        e.y = ny;
        log(state, `the ${e.kind} moves closer`);
        break;
      }
    }
  }
  state.turn = (state.turn || 0) + 1;
  state.actions_this_turn = 0;
  // A turn where nothing ran never reaches beginAction's clear, so cap the
  // log rather than let a run of failed turns accumulate one.
  if (state.log && state.log.length > 12) {
    state.log = state.log.slice(-12);
  }
  return state;
}

module.exports = { move, attack, pick_up, use_item, interact, end_turn };
